Struct mech/add patch test mpc

yamanatoo opened the pull request:
add the patch test using multi point constraint (MPC)

yamanatoo committed:
Patch Test MPC Plate Tension

yamanatoo committed:
changed order of InitializeResults and WriteNodalResults

yamanatoo committed:
added some checks

loumalouomega commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,164 @@
+    GidIO<> gid_io("TEST_MPC_Plate_Tension", GiD_PostBinary, SingleFile, WriteUndeformed,  WriteElementsOnly);
+    const int nl_iter = model_part.GetProcessInfo()[NL_ITERATION_NUMBER];
+    const double label = static_cast<double>(nl_iter);
+    gid_io.InitializeMesh(label);
```
Please comment GidIO and the std::cout

vryy replied:
did in https://github.com/yamanatoo/Kratos/commit/1a9d34e58552f8b2bb969235d8d4293833765575

loumalouomega commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,164 @@
+#include <iostream>
```
The header is missing

vryy replied:
i did not get what you mean?

loumalouomega replied:
~~~cpp
// KRATOS  ___|  |                   |                   |
//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
//             | |   |    |   | (    |   |   | |   (   | |
//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
//
//  License:             BSD License
//                                       license: StructuralMechanicsApplication/license.txt
//
//  Main authors:    YOUR NAME HERE
//
~~~

vryy replied:
done in 31b5debeb4808262370ed9575b2059030b8b639f

loumalouomega reviewed (changes requested):
Please comment debug info

vryy commented:
Hi, we make a patch test involved 3 elements and a hanging constraint. The test works as expected with the results as in pictures below.

![Displacement_X](https://user-images.githubusercontent.com/13781578/56144644-0afa6200-5fa3-11e9-8565-756aa5f876a0.png)
![Displacement_Y](https://user-images.githubusercontent.com/13781578/56144647-0cc42580-5fa3-11e9-8996-50794bfccf14.png)
![Stress_XX](https://user-images.githubusercontent.com/13781578/56144650-0e8de900-5fa3-11e9-9f1d-8a73b774dd6c.png)

yamanatoo committed:
Commented Gid output and results out

yamanatoo committed:
add header

loumalouomega reviewed (dismissed):
Thanks

loumalouomega commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,175 @@
+//
+
+#include <iostream>
+#include "includes/gid_io.h"
```
Can you comment this too?, will reduce the compilation time

vryy replied:
we will do this after the server test is successful. There are some other commits preventing this one to merge.

loumalouomega replied:
Yes, looks like is failing:

~~~sh
/home/travis/build/KratosMultiphysics/Kratos/applications//StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py
Check failed in file /usr/include/boost/numeric/ublas/storage.hpp at line 195:
i < size_
terminate called after throwing an instance of 'boost::numeric::ublas::bad_index'
  what():  bad index
~~~

vryy replied:
@yamanatoo can you trace down which line create it?

yamanatoo replied:
done in c92174185cb83cfc3baf170babeb4a740567bc1c

loumalouomega reviewed (commented):
Minor request

loumalouomega committed:
Minor

vryy commented:
it seems the server test failed somewhere with MeshMovingApplication. We will merge again when it passed.

philbucher reviewed (changes requested):
I also want to review this, wont take long, sorry for the inconvenience

philbucher commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,178 @@
+#include "includes/model_part.h"
+#include "spaces/ublas_space.h"
+#include "includes/linear_master_slave_constraint.h"
+#include "structural_mechanics_application.h"
```
```suggestion

```
this should not be included anywhere

philbucher commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,178 @@
+    model_part.AddNodalSolutionStepVariable(DISPLACEMENT);
+    model_part.AddNodalSolutionStepVariable(REACTION);
+    //Material properties
+    auto p_Properties = model_part.CreateNewProperties(0);
```
+1 for using new syntax

philbucher commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,178 @@
+    VariableUtils().AddDofWithReaction(DISPLACEMENT_Y, REACTION_Y, model_part);
+    // Initialize elements
+    auto& r_process_info = model_part.GetProcessInfo();
+    for (auto& elem : model_part.Elements())
```
I don't think this is needed since you call `Solve` on the Strategy, can you please test try if it works without?

philbucher commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,178 @@
+    // Set initial solution
+    for (auto& node : model_part.Nodes())
+    {
+        (node.FastGetSolutionStepValue(DISPLACEMENT)).clear();
```
are you sure this is needed?

philbucher commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,178 @@
+    Vector constant_vector = ZeroVector(1);
+    model_part.CreateNewMasterSlaveConstraint("LinearMasterSlaveConstraint", 1, master_dofs, slave_dofs, relation_matrix, constant_vector);
+    //Setting up System
+    SchemeType::Pointer p_scheme = SchemeType::Pointer( new ResidualBasedIncrementalUpdateStaticSchemeType() );
```
please use `Kratos::make_shared` instead of `new`

vryy replied:
what is the benefit of Kratos::make_shared? I guess to avoid a copy?

philbucher replied:
yes in a nutshell :)
https://stackoverflow.com/questions/18301511/stdshared-ptr-initialization-make-sharedfoo-vs-shared-ptrtnew-foo

also because we have widely adapted this style now

philbucher commented on applications/StructuralMechanicsApplication/tests/cpp_tests/test_mpc_plate_tension.cpp:
```diff
@@ -0,0 +1,178 @@
+// Linear solvers
+#include "linear_solvers/reorderer.h"
+#include "linear_solvers/direct_solver.h"
+#include "linear_solvers/linear_solver.h"
```
please clean your includes, there are a few you don't need

yamanatoo replied:
Thank you for the review and information
all above mentioned issues have been solved in c92174185cb83cfc3baf170babeb4a740567bc1c

vryy replied:
thanks @yamanatoo for your contribution

philbucher reviewed (commented):
as @loumalouomega points out the test is failing
=> I recommend to compile `FullDebug`

yamanatoo committed:
removed unused includes
removed unused typedefs
changed to (Kratos::makeshared) instead of (new)
included line_load_condition_2d.h instead of structural_mechanics_application.h
fixed constant_vector size

philbucher committed:
more includes removed

philbucher reviewed (approved):
thanks
